docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+{
+    Nullability ReturnNullability { get; }
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
+}
+
+public interface IParameterSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
+}
+
+public interface INamedTypeSymbol : ITypeSymbol
+{
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+}
+
+public interface ITypeParameterSymbol : ITypeSymbol
+{
+    Nullability ReferenceTypeConstraintNullability { get; }
+    ImmutableArray<Nullability> ConstraintsNullabilities { get; }
+}
+
+public struct TypeInfo
+{
+    public Nullability Nullability { get; }
+    public Nullability ConvertedNullability { get; }
+}
+
+#endregion
+
+#region Compilation
+
+public class Compilation
+{
+    public SemanticModel GetSemanticModel(bool skipNullabilityInformation);
+}
+
+
+#endregion
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
+
+#endregion
+```
+
+## Considered Alternatives
+
+### `Nullability` on `ITypeSymbol`
+
+The basic principle of this alternative proposal is to add `Nullability Nullability { get; }` to `ITypeSymbol`, and remove the rest of the proposed nullability retrieval APIs in the main section.
+This has a few benefits:
+
+1. Consumers of `ITypeSymbol`s won't have to keep track of where the type symbol came from to retrieve nullability information, they just have to ask the type.
+2. Printing APIs and code generation APIs do not need to have new overloads introduced to allow communication of top-level nullability information.
+3. Consumers of the new API don't have to separately worry about nullability when comparing types, as `string`, `string~`, and `string?` are truly different types.
+
+However, this approach has a few major issues that ultimately break our ability to do this design.
+
+1. This approach will violate principle #1 of our goals for this API, breaking existing analyzers and refactorings extensively.
+A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.
```

Comment:
Can you explain this.  Why would an existing analyzer/refactoring break?  

> A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.

These would not break.  Those fields are not of that type.   If i had a field of type 'string' before then it will still be found.  But if i change my API to use `string?` then the analyzer should *not* find it since htat's a different type, and there is no reason to believe that analyzer/refactoring will work on that type properly.

Reply:
But, to do that, it would need to do .OriginalDefinition anyways.  Because `Task<T>` won't match `Task<string>` or `Task<int>` etc. etc.  So it already needs to deal with variadic concept of type matching.